### PR TITLE
revise error messages with regards to DerivationIndex

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

ADP-509

# Overview

- 0d1f207902a45a5451a38b264156f92f17d596db
  :round_pushpin: **revise error messages and error code for DerivationIndex**
    The current errors were saying black and white at the same time. Understanding what was a valid a value for derivation index and what was not was unclear.

- 96c62fbb9b9380e8c7356730dd2b06f172c7e561
  :round_pushpin: **remove 409 Conflict from possible errors of the new signature endpoint.**

# Comments

### CODE REVIEW

**Review by piotr-iohk — APPROVED** [review]
LGTM.

File: lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs [hunk 1]
```diff
@@ -269,27 +269,29 @@ instance FromText DerivationIndex where
...
         firstHardened = getIndex @'Hardened minBound
 
+        errMalformed = TextDecodingError $ unwords
+            [ "A derivation index must be a natural number between"
+            , show (getIndex @'Soft minBound)
+            , "an "
+            , show (getIndex @'Soft maxBound)
+            , "with an optional 'H' suffix (e.g. \"1815H\" or \"44\")."
+            , "Indexes without suffix are called 'Soft'"
+            , "Indexes with suffix are called 'Hardened'."
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs]:
```suggestion
            , "Indexes without suffixes are called 'Soft'"
            , "Indexes with suffixes are called 'Hardened'."
```

File: lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs [hunk 2]
```diff
@@ -214,17 +214,17 @@ instance Malformed (PathParam (ApiT DerivationIndex)) where
...
-            "expected a number as string with an optional 'H' suffix \
-            \(e.g. '1815H' or '44')."
-
-        msgOutOfBounds =
-            "A derivation index must be a natural number between 0 and 2147483647."
+            "A derivation index must be a natural number between \
+            \0 and 2147483647 with an optional 'H' suffix \
+            \(e.g. \"1815H\" or \"44\"). \
+            \Indexes without suffix are called 'Soft' \
+            \Indexes with suffix are called 'Hardened'."
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs]:
```suggestion
            \Indexes without suffixes are called 'Soft' \
            \Indexes with suffixes are called 'Hardened'."
```

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 3]
```diff
@@ -471,21 +471,33 @@ spec = do
...
-                  , "between 0 and 2147483647."
+            let message = unwords
+                  [ "Error in $:"
+                  , "A derivation index must be a natural number between"
+                  , show (getIndex @'Soft minBound)
+                  , "an "
+                  , show (getIndex @'Soft maxBound)
+                  , "with an optional 'H' suffix (e.g. \"1815H\" or \"44\")."
+                  , "Indexes without suffix are called 'Soft'"
+                  , "Indexes with suffix are called 'Hardened'."
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]:
```suggestion
                  , "Indexes without suffixes are called 'Soft'"
                  , "Indexes with suffixes are called 'Hardened'."
```

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 4]
```diff
@@ -471,21 +471,33 @@ spec = do
...
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]:
```suggestion
                  , "Indexes without suffixes are called 'Soft'"
                  , "Indexes with suffixes are called 'Hardened'."
```

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 5]
```diff
@@ -2764,10 +2764,11 @@ instance LiftHandler ErrDerivePublicKey where
...
     handler = \case
         ErrIndexTooHigh maxIx _ix ->
-            apiError err400 BadRequest $ mconcat
+            apiError err403 SoftDerivationRequired $ mconcat
                 [ "It looks like you've provided a derivation index that is "
-                , "out of bound. I can only derive keys up to #"
-                , pretty maxIx, "."
+                , "out of bound. The index is well-formed, but I require "
+                , "indexes valid for soft derivation only. That is, indexes "
+                , "between 0 and ", pretty maxIx, " without suffix."
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]:
```suggestion
                , "between 0 and ", pretty maxIx, " without a suffix."
```

**Comment by KtorZ** [comment]
bors merge

**Review by paweljakubas — APPROVED** [review]

**Pull request merged**